File: TP4/src/java/TP4.java

// #region IMPORTS
import java.io.*;
import java.util.*;
import java.time.*;
// #endregion

interface AED {

  public void inserir(Player pos);

  public void remover(String pos);

  public boolean search(String toSearch);

  public boolean searchPrinting(String toSearch);

  public void mostrar();
}

public class TP4 {
  public static void main(String[] args) {
    int EXERCISE_NUMBER = 2;
    AED structure = GetStructureBasedOnExerciseNumber(EXERCISE_NUMBER);
    boolean controller = false;
    String word = "";
    Scanner reader = new Scanner(System.in);
    do {
      word = reader.nextLine();
      controller = IsAble(word);
      if (!controller) {
        Player util = CreateNewPlayerFromFile(word);
        // if (!structure.search(util.getName()))
        structure.inserir(util);
      }
    } while (!controller);
    boolean resp = false;
    do {
      word = reader.nextLine();
      controller = IsAble(word);
      if (!controller) {
        if (EXERCISE_NUMBER == 5) {
          if (!structure.search(word)) {
            structure.search(word);
          }
        } else {
          System.out.print(word + " raiz");
          resp = structure.searchPrinting(word);
          if (!resp)
            System.out.println(" NAO");
          else
            System.out.println(" SIM");
        }
      }
    } while (!controller);
    if (EXERCISE_NUMBER == 5) {
      structure.mostrar();
    }
    reader.close();
  }

  public static AED GetStructureBasedOnExerciseNumber(int exercise) {
    AED valueToReturn;
    if (exercise == 1 || exercise == 5) {
      valueToReturn = new ArvoreBinaria();
    } else if (exercise == 4) {
      valueToReturn = new Alvinegra();
    } else if (exercise == 2) {
      valueToReturn = new BinariaDeBinaria();
    } else {
      valueToReturn = new Trie();
    }
    return valueToReturn;
  }

  public static Player CreateNewPlayerFromFile(String valueToRead) {
    Player valueToReturn = null;
    try {
      Scanner csvReader = new Scanner(new File("/tmp/players.csv"));
      int lineToread = Integer.parseInt(valueToRead);
      String auxWord = "";
      for (int i = 0; i < lineToread + 2; i++) {
        auxWord = csvReader.nextLine();
      }
      csvReader.close();
      String[] splitted = auxWord.split(",");
      splitted = FixEntry(splitted);
      valueToReturn = new Player(splitted);
    } catch (Exception e) {
      System.out.println("Erro ao criar o jogador: ");
    }
    return valueToReturn;
  }

  public static boolean IsAble(String _word) {
    return _word.charAt(0) == ('F') && _word.charAt(1) == ('I') && _word.charAt(2) == ('M');
  }

  public static String[] FixEntry(String[] word) {
    if (word.length != 8) {
      String[] valueToReturn = new String[8];
      String NULO = "nao informado";
      for (int i = 0; i < word.length; i++) {
        if (!word[i].equals(""))
          valueToReturn[i] = word[i];
        else
          valueToReturn[i] = NULO;
      }
      if (word.length != 8) {
        for (int i = word.length; i < valueToReturn.length; i++) {
          valueToReturn[i] = NULO;
        }
      }
      return valueToReturn;
    } else {
      return word;
    }
  }

}

class Player {
  private int id, heigth, weigth;
  private String name, university, birthYear, birthCity, birthState;
  private final String NULO = "nao informado";

  // #region <----------- CTOR & GETTERS & SETTERS ------------------>
  public Player(int id, int heigth, int weigth, String name, String university, String birthYear, String birthCity,
      String birthState) {
    this.id = id;
    this.heigth = heigth;
    this.weigth = weigth;
    this.name = name;
    this.university = university;
    this.birthYear = birthYear;
    this.birthCity = birthCity;
    this.birthState = birthState;
  }

  public Player(String[] splitted) {
    setId(Integer.parseInt(splitted[0]));
    setName(splitted[1]);
    setHeigth(Integer.parseInt(splitted[2]));
    setWeigth(Integer.parseInt(splitted[3]));
    setUniversity(splitted[4]);
    setBirthYear(splitted[5]);
    setBirthCity(splitted[6]);
    setBirthState(splitted[7]);
  }

  public int getId() {
    return id;
  }

  public void setId(int id) {
    this.id = id;
  }

  public int getHeigth() {
    return heigth;
  }

  public void setHeigth(int heigth) {
    this.heigth = heigth;
  }

  public int getWeigth() {
    return weigth;
  }

  public void setWeigth(int weigth) {
    this.weigth = weigth;
  }

  public String getName() {
    return name;
  }

  public void setName(String name) {
    this.name = (!name.equals("")) ? name : NULO;
  }

  public String getUniversity() {
    return university;
  }

  public void setUniversity(String university) {
    this.university = (!university.equals("")) ? university : NULO;
  }

  public String getBirthYear() {
    return birthYear;
  }

  public void setBirthYear(String birthYear) {
    this.birthYear = (!birthYear.equals("")) ? birthYear : NULO;
  }

  public String getBirthCity() {
    return birthCity;
  }

  public void setBirthCity(String birthCity) {
    this.birthCity = (!birthCity.equals("")) ? birthCity : NULO;
  }

  public String getBirthState() {
    return birthState;
  }

  public void setBirthState(String birthState) {
    this.birthState = (!birthState.equals("")) ? birthState : NULO;
  }
  // #endregion

  public void PrintPlayer() {
    StringBuilder stringBuilder = new StringBuilder();
    // stringBuilder.append(" ## ");
    stringBuilder.append(getName());
    // stringBuilder.append(" ## ");
    // stringBuilder.append(getHeigth());
    // stringBuilder.append(" ## ");
    // stringBuilder.append(getWeigth());
    // stringBuilder.append(" ## ");
    // stringBuilder.append(getBirthYear());
    // stringBuilder.append(" ## ");
    // stringBuilder.append(getUniversity());
    // stringBuilder.append(" ## ");
    // stringBuilder.append(getBirthCity());
    // stringBuilder.append(" ## ");
    // stringBuilder.append(getBirthState());
    // stringBuilder.append(" ## ");
    System.out.println(stringBuilder.toString());
  }

}

// #region DATASTRUCTURES

// #region NODES

class No {
  public Player elemento;
  public No esq, dir;

  public No(Player elemento) {
    this(elemento, null, null);
  }

  public No(Player elemento, No esq, No dir) {
    this.elemento = elemento;
    this.esq = esq;
    this.dir = dir;
  }
}

class BlackNo {
  ArvoreBinaria element;
  int height;
  BlackNo left, right;

  public BlackNo(int _h) {
    height = _h;
    element = new ArvoreBinaria();
    left = right = null;
  }
}

class NoAN {
  public boolean cor;
  public Player elemento;
  public NoAN esq, dir;

  public NoAN(Player elemento) {
    this(elemento, false, null, null);
  }

  public NoAN(Player elemento, boolean cor) {
    this(elemento, cor, null, null);
  }

  public NoAN(Player elemento, boolean cor, NoAN esq, NoAN dir) {
    this.cor = cor;
    this.elemento = elemento;
    this.esq = esq;
    this.dir = dir;
  }
}

// #endregion

// #region BST

class ArvoreBinaria implements AED {
  public No raiz; // Raiz da arvore.

  public ArvoreBinaria() {
    raiz = null;
  }

  @Override
  public boolean searchPrinting(String x) {
    return searchPrintingRec(x, raiz);
  }

  private boolean searchPrintingRec(String x, No i) {
    boolean resp = false;
    if (i != null) {
      resp = x.compareTo(i.elemento.getName()) == 0;
      if (!resp) {
        System.out.print(" esq");
        if (searchPrintingRec(x, i.esq)) {
          resp = true;
        } else {
          System.out.print(" dir");
          if (searchPrintingRec(x, i.dir)) {
            resp = true;
          }
        }
      }
    }
    return resp;
  }

  @Override
  public boolean search(String x) {
    return searchRec(x, raiz);
  }

  private boolean searchRec(String x, No i) {
    boolean resp = false;
    if (i != null) {
      int comparison = x.compareTo(i.elemento.getName());
      if (0 == comparison) {
        resp = true;
      } else if (0 > comparison) {
        resp = searchRec(x, i.esq);
      } else {
        resp = searchRec(x, i.dir);
      }
    }
    return resp;
  }

  public void mostrar() {
    caminharCentral(raiz);
  }

  private void caminharCentral(No i) {
    if (i != null) {
      caminharCentral(i.esq); // Elementos da esquerda.
      i.elemento.PrintPlayer();
      caminharCentral(i.dir); // Elementos da direita.
    }
  }

  public void inserir(Player x) {
    raiz = inserir(x, raiz);
  }

  private No inserir(Player x, No i) {
    int comparison = 0;
    if (i == null) {
      i = new No(x);

    } else {
      comparison = x.getName().compareTo(i.elemento.getName());
    }
    if (0 > comparison) {
      i.esq = inserir(x, i.esq);
    } else if (0 < comparison) {
      i.dir = inserir(x, i.dir);
    }
    return i;
  }

  public void remover(String x) {
    raiz = remover(x, raiz);
  }

  private No remover(String x, No i) {
    int comparison = 0;
    if (i == null) {
      System.out.println("ERRO AO REMOVER");
    } else {
      comparison = x.compareTo(i.elemento.getName());
    }
    if (0 > comparison) {
      i.esq = remover(x, i.esq);
    } else if (0 < comparison) {
      i.dir = remover(x, i.dir);
    } else if (i.dir == null) {
      i = i.esq;
    } else if (i.esq == null) {
      i = i.dir;
    } else {
      i.esq = antecessor(i, i.esq);
    }
    return i;
  }

  private No antecessor(No i, No j) {
    if (j.dir != null) {
      j.dir = antecessor(i, j.dir);
    } else {
      i.elemento = j.elemento;
      j = j.esq;
    }
    return j;
  }

}

// #endregion

// #region ALVINEGRA
class Alvinegra implements AED {
  public NoAN raiz; // Raiz da arvore.

  public Alvinegra() {
    raiz = null;
  }

  public void inserir(Player elemento) {
    if (raiz == null) {
      raiz = new NoAN(elemento, false);
    } else if (raiz.esq == null && raiz.dir == null) {
      if (raiz.elemento.getName().compareTo(elemento.getName()) > 0) {
        raiz.esq = new NoAN(elemento, true);
      } else {
        raiz.dir = new NoAN(elemento, true);
      }
    } else if (raiz.esq == null) {
      if (raiz.elemento.getName().compareTo(elemento.getName()) > 0) {
        raiz.esq = new NoAN(elemento);
      } else if (raiz.dir.elemento.getName().compareTo(elemento.getName()) > 0) {
        raiz.esq = new NoAN(raiz.elemento);
        raiz.elemento = elemento;
      } else {
        raiz.esq = new NoAN(raiz.elemento);
        raiz.elemento = raiz.dir.elemento;
        raiz.dir.elemento = elemento;
      }
      raiz.esq.cor = raiz.dir.cor = false;
    } else if (raiz.dir == null) {
      if (raiz.elemento.getName().compareTo(elemento.getName()) < 0) {
        raiz.dir = new NoAN(elemento);
      } else if (raiz.esq.elemento.getName().compareTo(elemento.getName()) < 0) {
        raiz.dir = new NoAN(raiz.elemento);
        raiz.elemento = elemento;
      } else {
        raiz.dir = new NoAN(raiz.elemento);
        raiz.elemento = raiz.esq.elemento;
        raiz.esq.elemento = elemento;
      }
      raiz.esq.cor = raiz.dir.cor = false;
    } else {
      insert(elemento, null, null, null, raiz);
    }
    raiz.cor = false;
  }

  private void insert(Player elemento, NoAN bisavo, NoAN avo, NoAN pai, NoAN i) {
    if (i == null) {
      if (elemento.getName().compareTo(pai.elemento.getName()) < 0) {
        i = pai.esq = new NoAN(elemento, true);
      } else {
        i = pai.dir = new NoAN(elemento, true);
      }
      if (pai.cor == true) {
        balancear(bisavo, avo, pai, i);
      }
    } else {
      if (i.esq != null && i.dir != null && i.esq.cor == true && i.dir.cor == true) {
        i.cor = true;
        i.esq.cor = i.dir.cor = false;
        if (i == raiz) {
          i.cor = false;
        } else if (pai.cor == true) {
          balancear(bisavo, avo, pai, i);
        }
      }
      if (elemento.getName().compareTo(i.elemento.getName()) < 0) {
        insert(elemento, avo, pai, i, i.esq);
      } else if (elemento.getName().compareTo(i.elemento.getName()) > 0) {
        insert(elemento, avo, pai, i, i.dir);
      } else {
        System.out.println("Erro inserir (elemento repetido)!");
      }
    }
  }

  private void balancear(NoAN bisavo, NoAN avo, NoAN pai, NoAN i) {
    if (pai.cor == true) {
      if (pai.elemento.getName().compareTo(avo.elemento.getName()) > 0) {
        if (i.elemento.getName().compareTo(pai.elemento.getName()) > 0) {
          avo = rotacaoEsq(avo);
        } else {
          avo = rotacaoDirEsq(avo);
        }
      } else {
        if (i.elemento.getName().compareTo(pai.elemento.getName()) < 0) {
          avo = rotacaoDir(avo);
        } else {
          avo = rotacaoEsqDir(avo);
        }
      }
      if (bisavo == null) {
        raiz = avo;
      } else {
        if (avo.elemento.getName().compareTo(bisavo.elemento.getName()) < 0) {
          bisavo.esq = avo;
        } else {
          bisavo.dir = avo;
        }
      }
      avo.cor = false;
      avo.esq.cor = avo.dir.cor = true;
    }
  }

  private NoAN rotacaoDir(NoAN no) {
    NoAN noEsq = no.esq;
    NoAN noEsqDir = noEsq.dir;
    noEsq.dir = no;
    no.esq = noEsqDir;
    return noEsq;
  }

  private NoAN rotacaoEsq(NoAN no) {
    NoAN noDir = no.dir;
    NoAN noDirEsq = noDir.esq;
    noDir.esq = no;
    no.dir = noDirEsq;
    return noDir;
  }

  private NoAN rotacaoDirEsq(NoAN no) {
    no.dir = rotacaoDir(no.dir);
    return rotacaoEsq(no);
  }

  private NoAN rotacaoEsqDir(NoAN no) {
    no.esq = rotacaoEsq(no.esq);
    return rotacaoDir(no);
  }

  @Override
  public boolean search(String toSearch) {
    return pesquisarRec(toSearch, raiz);
  }

  private boolean pesquisarRec(String elemento, NoAN i) {
    boolean resp = false;
    int comparison = 0;
    if (i == null) {
      resp = false;
    } else {
      comparison = elemento.compareTo(i.elemento.getName());
    }
    if (i != null && comparison == 0) {
      resp = true;
    } else if (comparison < 0) {
      resp = pesquisarRec(elemento, i.esq);
    } else {
      try {
        resp = pesquisarRec(elemento, i.dir);
      } catch (Exception e) {
      }
    }
    return resp;
  }

  @Override
  public boolean searchPrinting(String toSearch) {
    return searchPrintingRec(toSearch, raiz);
  }

  private boolean searchPrintingRec(String elemento, NoAN i) {
    boolean resp = false;
    int comparison;
    if (i != null) {
      comparison = elemento.compareTo(i.elemento.getName());
      if (comparison == 0) {
        resp = true;
      } else if (comparison < 0) {
        System.out.print(" esq");
        resp = searchPrintingRec(elemento, i.esq);
      } else {
        try {
          System.out.print(" dir");
          resp = searchPrintingRec(elemento, i.dir);
        } catch (Exception e) {
        }
      }
    }
    return resp;
  }

  @Override
  public void mostrar() {
    // TODO Auto-generated method stub
  }

  @Override
  public void remover(String pos) {
    // TODO Auto-generated method stub

  }
}

// #endregion

// #region BINARIA DE BINARIA
class BinariaDeBinaria implements AED {

  public BlackNo raiz;

  public BinariaDeBinaria() {
    raiz = new BlackNo(7);
    raiz.right = new BlackNo(11);
    raiz.left = new BlackNo(3);
    raiz.right.right = new BlackNo(13);
    raiz.right.left = new BlackNo(9);
    raiz.left.right = new BlackNo(5);
    raiz.left.left = new BlackNo(1);
    raiz.left.left.right = new BlackNo(2);
    raiz.left.left.left = new BlackNo(0);
    raiz.left.right.right = new BlackNo(6);
    raiz.left.right.left = new BlackNo(4);
    raiz.right.left.right = new BlackNo(10);
    raiz.right.left.left = new BlackNo(8);
    raiz.right.right.right = new BlackNo(14);
    raiz.right.right.left = new BlackNo(12);
  }

  public BinariaDeBinaria(int n) {
    raiz = new BlackNo(n);
  }

  @Override
  public void inserir(Player pos) {
    int calc = pos.getHeigth() % 15;
    inserirRec(pos, raiz, calc);
  }

  public void inserirRec(Player p, BlackNo n, int calc) {
    if (n.height == calc) {
      n.element.inserir(p);
    } else if (n.height < calc) {
      inserirRec(p, n.right, calc);
    } else {
      inserirRec(p, n.left, calc);
    }
  }

  @Override
  public boolean search(String toSearch) {
    return searchRec(toSearch, raiz);
  }

  public boolean searchRec(String name, BlackNo i) {
    boolean response = false;
    if (i != null) {
      if (i.element.search(name)) {
        response = true;
      }
      if (!response) {
        if (searchRec(name, i.left)) {
          response = true;
        }
      }
      if (!response) {
        if (searchRec(name, i.right)) {
          response = true;
        }
      }
    }
    return response;
  }

  @Override
  public boolean searchPrinting(String toSearch) {
    return searchingPrinting(toSearch, raiz);
  }

  public boolean searchingPrinting(String name, BlackNo i) {
    boolean response = false;
    if (i != null) {
      response = (i.element.searchPrinting(name));
      if (!response) {
        System.out.print(" ESQ");
        if (searchingPrinting(name, i.left)) {
          response = true;
        } else {
          System.out.print(" DIR");
          if (searchingPrinting(name, i.right)) {
            response = true;
          }
        }
      }
    }
    return response;
  }

  @Override
  public void mostrar() {
    // TODO Auto-generated method stub

  }

  @Override
  public void remover(String pos) {
    // TODO Auto-generated method stub

  }
}

// #endregion

// #region TRIE
class Trie implements AED {

  @Override
  public void inserir(Player pos) {
    // TODO Auto-generated method stub

  }

  @Override
  public boolean search(String toSearch) {
    // TODO Auto-generated method stub
    return false;
  }

  @Override
  public boolean searchPrinting(String toSearch) {
    // TODO Auto-generated method stub
    return false;
  }

  @Override
  public void mostrar() {
    // TODO Auto-generated method stub

  }

  @Override
  public void remover(String pos) {
    // TODO Auto-generated method stub

  }
}

// #endregion

// #endregion

File: TP4/src/c/TP4.cpp
// < ------------ HEADER FILES   ------------ >

#include <math.h>
#include <time.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <stdbool.h>
#include <err.h>

// < ------------ DECLARATION METHODS   ------------ >

typedef struct Players
{
  int id;
  char *name;
  int heigth;
  int weigth;
  char *university;
  char *birthYear;
  char *birthCity;
  char *birthState;
} Player;

typedef struct Node
{
  Player data;        //element in Node
  int ht;             //height of this Node
  struct Node *left;  //left Node
  struct Node *right; // Right Node
} node;

// < ------------ UTILS   ------------ >

char *Split(char **word)
{
  return strsep(word, " ");
}
bool Equals(char *first, char *second)
{
  return strcmp(first, second) == 0;
}

// < ------------ DATA STRUCTURES   ------------ >

char *Replace(char const *const original, char const *const pattern, char const *const replacement)
{
  size_t const replen = strlen(replacement);
  size_t const patlen = strlen(pattern);
  size_t const orilen = strlen(original);
  size_t patcnt = 0;
  const char *oriptr;
  const char *patloc;
  for (oriptr = original; patloc = strstr(oriptr, pattern); oriptr = patloc + patlen)
  {
    patcnt++;
  }
  {
    size_t const retlen = orilen + patcnt * (replen - patlen);
    char *const returned = (char *)malloc(sizeof(char) * (retlen + 1));
    if (returned != NULL)
    {
      char *retptr = returned;
      for (oriptr = original; patloc = strstr(oriptr, pattern); oriptr = patloc + patlen)
      {
        size_t const skplen = patloc - oriptr;
        strncpy(retptr, oriptr, skplen);
        retptr += skplen;
        strncpy(retptr, replacement, replen);
        retptr += replen;
      }
      strcpy(retptr, oriptr);
    }
    return returned;
  }
}

void PrintPlayer(Player *player)
{
  printf("## %s ## %d ## %d ## %s ## %s ## %s ## %s ##\n", player->name, player->heigth, player->weigth, player->birthYear, player->university, player->birthCity, player->birthState);
}

void PlayerSet(int checker, Player *playerToReturn, char *field)
{
  if (checker == 0)
  {
    playerToReturn->id = atoi(field);
  }
  else if (checker == 1)
  {
    playerToReturn->name = field;
  }
  else if (checker == 2)
  {
    playerToReturn->heigth = atoi(field);
  }
  else if (checker == 3)
  {
    playerToReturn->weigth = atoi(field);
  }
  else if (checker == 4)
  {
    playerToReturn->university = field;
  }
  else if (checker == 5)
  {
    playerToReturn->birthYear = field;
  }
  else if (checker == 6)
  {
    //FIX \n at the end
    int place = strlen(field);
    if (field[place - 1] == '\n')
    {
      field[place - 1] = field[place];
      // printf("%s\n", field);
    }
    playerToReturn->birthCity = field;
  }
  else if (checker == 7)
  {
    int place = strlen(field);
    if (field[place - 1] == '\n')
    {
      field[place - 1] = field[place];
    }
    playerToReturn->birthState = field;
  }
}

Player GetPlayerFromFile(char *word)
{
  // printf("%s\n", word);
  Player playerToReturn;
  char *field = strtok(word, ",");
  int checker = 0;
  while (field)
  {
    PlayerSet(checker, &playerToReturn, field);
    checker++;
    field = strtok(NULL, ",");
  }
  return playerToReturn;
}

char *GetLineFromFile(int lineToRead)
{
  FILE *csvReader = fopen("/tmp/players.csv", "r");
  static char buffer[1024];
  fgets(buffer, 1024, csvReader);
  for (int i = 0; i < lineToRead; i++)
  {
    fgets(buffer, 1024, csvReader);
  }
  fgets(buffer, 1024, csvReader);
  fclose(csvReader);
  return buffer;
}

Player GetPlayerFromLine(char *player)
{
  char *buffer;
  buffer = GetLineFromFile(atoi(player)); //blocao contendo todos os dados da linha que representa um jogador
  char *test = Replace(buffer, ",,", ",nao informado,");
  char *final = Replace(test, ",\n", ",nao informado\n");
  free(test);
  return GetPlayerFromFile(final);
}

bool IsAble(char *_word)
{
  bool valueToReturn = false;
  if (_word[0] == ('F') && _word[1] == ('I') && _word[2] == ('M'))
  {
    valueToReturn = true;
  }
  return valueToReturn;
}

int height(node *T)
{
  int lh, rh;
  if (T == NULL)
    return (0);
  if (T->left == NULL)
    lh = 0;
  else
    lh = 1 + T->left->ht;
  if (T->right == NULL)
    rh = 0;
  else
    rh = 1 + T->right->ht;
  if (lh > rh)
    return (lh);
  return (rh);
}

int BF(node *T)
{
  int lh, rh;
  if (T == NULL)
    return (0);
  if (T->left == NULL)
    lh = 0;
  else
    lh = 1 + T->left->ht;
  if (T->right == NULL)
    rh = 0;
  else
    rh = 1 + T->right->ht;
  return (lh - rh);
}

node *rotateright(node *x)
{
  node *y;
  y = x->left;
  x->left = y->right;
  y->right = x;
  x->ht = height(x);
  y->ht = height(y);
  return (y);
}

node *rotateleft(node *x)
{
  node *y;
  y = x->right;
  x->right = y->left;
  y->left = x;
  x->ht = height(x);
  y->ht = height(y);
  return (y);
}

node *RR(node *T)
{
  T = rotateleft(T);
  return (T);
}

node *LL(node *T)
{
  T = rotateright(T);
  return (T);
}

node *LR(node *T)
{
  T->left = rotateleft(T->left);
  T = rotateright(T);
  return (T);
}

node *RL(node *T)
{
  T->right = rotateright(T->right);
  T = rotateleft(T);
  return (T);
}

void preorder(node *T)
{
  if (T != NULL)
  {
    PrintPlayer(&T->data);
    preorder(T->left);
    preorder(T->right);
  }
}

node *insert(node *T, Player x)
{
  if (T == NULL)
  {
    T = (node *)malloc(sizeof(node));
    T->data = x;
    T->left = NULL;
    T->right = NULL;
  }
  else if (strcmp(x.name, T->data.name) > 0) // insert in right subtree
  {
    T->right = insert(T->right, x);
    if (BF(T) == -2)
      if (strcmp(x.name, T->right->data.name) > 0)
        T = RR(T);
      else
        T = RL(T);
  }
  else if (strcmp(x.name, T->data.name) < 0) // insert in right subtree
  {
    T->left = insert(T->left, x);
    if (BF(T) == 2)
      if (strcmp(x.name, T->left->data.name) < 0)
        T = LL(T);
      else
        T = LR(T);
  }
  T->ht = height(T);
  return (T);
}

bool search(node *T, char *str)
{
  bool response = false;
  if (T != NULL)
  {
    int comparisonValue = strcmp(str, T->data.name);
    if (comparisonValue == 0)
    {
      response = true;
    }
    else if (comparisonValue == -1)
    {
      response = search(T->left, str);
    }
    else if (comparisonValue == 1)
    {
      response = search(T->right, str);
    }
  }
  return response;
}

bool searchPrinting(node *T, char *str)
{
  bool response = false;
  if (T != NULL)
  {
    int comparisonValue = strcmp(str, T->data.name);
    if (comparisonValue == 0)
    {
      response = true;
    }
    else if (comparisonValue < 0)
    {
      printf(" esq");
      response = searchPrinting(T->left, str);
    }
    else if (comparisonValue >0)
    {
      printf(" dir");
      response = searchPrinting(T->right, str);
    }
  }
  return response;
}

// < ------------ MAIN   ------------ >

int main()
{
  char word[1024];
  bool controller = false;
  scanf("%[^\n]%*c", word);
  node *root = NULL;
  do
  {
    controller = IsAble(word);
    if (!controller)
    {
      Player toInsert = GetPlayerFromLine(word);
      root = insert(root, toInsert);
    }
    scanf("%[^\n]%*c", word);
  } while (!controller);
  do
  {
    controller = IsAble(word);
    if (!controller)
    {
      printf("%s raiz", word);
      bool val = searchPrinting(root, word);
      if (val)
      {
        printf(" SIM");
      }
      else
      {
        printf(" NAO");
      }
      printf("\n");
    }
    scanf("%[^\n]%*c", word);
  } while (!controller);
  return 0;
}